packages/rpc-methods/src/restricted/manageAccounts.ts

```diff
@@ -0,0 +1,133 @@
+import {
+  SubjectType,
+  PermissionType,
+  RestrictedMethodOptions,
+  ValidPermissionSpecification,
+  PermissionSpecificationBuilder,
+} from '@metamask/permission-controller';
+import { Json, JsonStruct, NonEmptyArray } from '@metamask/utils';
+import { ethErrors } from 'eth-rpc-errors';
+import { assert, string, tuple } from 'superstruct';
+
+export const methodName = 'snap_manageAccounts';
+
+export type ManageAccountsMethodHooks = {
+  /**
+   * Gets the snap keyring implementation.
+   */
+  getSnapKeyring: (snapOrigin: string) => Promise<{
+    handleKeyringSnapMessage: (snapId: string, message: any) => Promise<Json>;
+  }>;
+
+  /**
+   * Saves the snap keyring, should be called after mutable operations.
+   *
+   * If an account was removed the address of the account must be passed to
+   * update the UI.
+   */
+  saveSnapKeyring: () => Promise<void>;
+};
+
+type ManageAccountsSpecificationBuilderOptions = {
+  allowedCaveats?: Readonly<NonEmptyArray<string>> | null;
+  methodHooks: ManageAccountsMethodHooks;
+};
+
+type ManageAccountsSpecification = ValidPermissionSpecification<{
+  permissionType: PermissionType.RestrictedMethod;
+  targetName: typeof methodName;
+  methodImplementation: ReturnType<typeof manageAccountsImplementation>;
+  allowedCaveats: Readonly<NonEmptyArray<string>> | null;
+}>;
+
+/**
+ * The specification builder for the `snap_manageAccounts` permission.
+ * `snap_manageAccounts` lets the Snap manage a set of accounts via a custom keyring.
+ *
+ * @param options - The specification builder options.
+ * @param options.allowedCaveats - The optional allowed caveats for the permission.
+ * @param options.methodHooks - The RPC method hooks needed by the method implementation.
+ * @returns The specification for the `snap_manageAccounts` permission.
+ */
+export const specificationBuilder: PermissionSpecificationBuilder<
+  PermissionType.RestrictedMethod,
+  ManageAccountsSpecificationBuilderOptions,
+  ManageAccountsSpecification
+> = ({
+  allowedCaveats = null,
+  methodHooks,
+}: ManageAccountsSpecificationBuilderOptions) => {
+  return {
+    permissionType: PermissionType.RestrictedMethod,
+    targetName: methodName,
+    allowedCaveats,
+    methodImplementation: manageAccountsImplementation(methodHooks),
+    subjectTypes: [SubjectType.Snap],
+  };
+};
+
+/**
+ * Validates the manageAccount method parameter `message`.
+ *
+ * @param message - Parameter to be validated.
+ */
+export function validateParams(message: unknown): void {
+  if (!Array.isArray(message)) {
+    throw ethErrors.rpc.invalidParams(
+      'Invalid ManageAccount Arguments: An array of type SnapMessage was expected',
+    );
+  }
+
+  const [method] = message;
```

Comment:
Can we use an object for the params, rather than an array? We've been changing the other JSON-RPC methods to do so too.

Reply:
Updated in [de277d2](https://github.com/MetaMask/snaps/pull/1290/commits/de277d26a7810d3336f9fb5c63f868bb001cd9a8)